controller/sync.go

```diff
@@ -603,17 +608,24 @@ func (sc *syncContext) terminate() {
 	sc.log.Debug("terminating")
 	tasks, _ := sc.getSyncTasks()
 	for _, task := range tasks {
-		if !task.isHook() || !task.completed() {
+		if !task.isHook() || task.liveObj == nil {
 			continue
 		}
-		if isRunnable(task.groupVersionKind()) {
+		phase, msg, err := sc.getOperationPhase(task.liveObj)
```

Comment:
Note that this change is significant, breaking change of a corner case. The new behavior is that we will delete any **_in-flight_** hook resource when an operation is terminated, where in-flight means `v1alpha1.HealthStatusProgressing` and `v1alpha1.HealthStatusSuspended`. This means that if a progressing Deployment, PVC, StatefulSet, ReplicaSet, Service, Ingress, APIService  is used as a sync hook, then those resources will become deleted when the operation is terminated.

That said:
* using things like Deployments was never the intended use case of hooks (since that use case was intention of waves)
*  the new behavior is probably the correct behavior, since without this change we cannot use things like other CRDs (TensorflowJob) as a hook.

We either need to list this as a breaking change, or we need to modify the code such that when it sees Deployment, PVC, StatefulSet, ReplicaSet, Service, Ingress, APIService as a Sync hook, it will choose not to delete the resource.

Reply:
👍  working on PR with docs changes